### PR TITLE
Add ssh signing algorithm as a role option.  

### PR DESCRIPTION
**Pull request by ncabatoff**
This is based on @mtorromeo's work in #8383.

### CODE REVIEW

**Comment by ncabatoff** [comment]
Fixes #8414.

File: builtin/logical/ssh/path_sign.go [hunk 1]
```diff
@@ -511,10 +512,29 @@ func (b *creationBundle) sign() (retCert *ssh.Certificate, retErr error) {
 		},
 	}
 
+	sshAlgorithmSigner, _ := b.Signer.(ssh.AlgorithmSigner)
+
+	// prepare certificate for signing
+	certificate.Nonce = make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, certificate.Nonce); err != nil {
+		return nil, fmt.Errorf("failed to generate signed SSH key")
```
Comment (alexanderbez) [inline builtin/logical/ssh/path_sign.go:520]: Do we not want to wrap the error here to provide additional context?

File: builtin/logical/ssh/path_sign.go [hunk 2]
```diff
@@ -511,10 +512,29 @@ func (b *creationBundle) sign() (retCert *ssh.Certificate, retErr error) {
...
+	}
+	certificate.SignatureKey = sshAlgorithmSigner.PublicKey()
+
+	// get bytes to sign
+	c2 := *certificate
+	c2.Signature = nil
+	out := c2.Marshal()
+	certificateBytes := out[:len(out)-4]
```
Comment (alexanderbez) [inline builtin/logical/ssh/path_sign.go:528]: Why are we stripping off the last 4 bytes? Maybe add a comment to clarify that?

**Comment by flowerysong** [comment]
Hmm, I wasn't expecting someone else to jump on this so quickly. https://github.com/flowerysong/vault/commit/c19016318586a723a0ecae4678e57dcf007e9b24 is what I have; the main difference is that it exposes it as an option at signing time, so you don't have to set up multiple roles if you need to support multiple signing algorithms.

File: builtin/logical/ssh/path_roles.go [hunk 3]
```diff
@@ -410,6 +431,7 @@ func (b *backend) pathRoleWrite(ctx context.Context, req *logical.Request, d *fr
 			KeyType:         KeyTypeOTP,
 			Port:            port,
 			AllowedUsers:    allowedUsers,
+			AlgorithmSigner: algorithmSigner,
```
Comment (flowerysong) [inline builtin/logical/ssh/path_roles.go:434]: This setting is only meaningful for KeyTypeCA, so you probably shouldn't add it to the other types.

File: builtin/logical/ssh/path_sign.go (hunk 2 above)
Comment (flowerysong) [reply builtin/logical/ssh/path_sign.go:528]: These bytes appear to hold the length of...something.

https://github.com/golang/crypto/blob/master/ssh/certs.go#L460

**Comment by mtorromeo** [comment]
> Hmm, I wasn't expecting someone else to jump on this so quickly. [flowerysong@c190163](https://github.com/flowerysong/vault/commit/c19016318586a723a0ecae4678e57dcf007e9b24) is what I have; the main difference is that it exposes it as an option at signing time, so you don't have to set up multiple roles if you need to support multiple signing algorithms.

Not having to duplicate roles is certainly preferable. This is not affecting me directly since I am fine using only sha2 but I see how it can be annoying.

**Comment by ncabatoff** [comment]
> Hmm, I wasn't expecting someone else to jump on this so quickly. [flowerysong@c190163](https://github.com/flowerysong/vault/commit/c19016318586a723a0ecae4678e57dcf007e9b24) is what I have; the main difference is that it exposes it as an option at signing time, so you don't have to set up multiple roles if you need to support multiple signing algorithms.

Hah!  Sorry about that @flowerysong.  I'm happy to go with your branch instead, you seem to know the backend better than I do anyway.  If you want to pursue it, be aware that the ssh acceptance tests (enabled by setting VAULT_ACC env var) appear to have been broken by a recent change.  Before we merge any PR, those should be fixed, and we should add tests for this change.  I was planning on also setting up a docker container running openssh to run the acceptance tests so we can run them in CI.  If you want to take that on, or you want to collaborate on these changes, let me know.  I'll hold off on my work in the meantime.

File: builtin/logical/ssh/path_roles.go (hunk 3 above)
Comment (ncabatoff) [reply builtin/logical/ssh/path_roles.go:434]: Good point.  I should've made this a draft PR, it's not yet been tested.  I wanted to put something up to show the previous PR creator what was needed next to make this configurable.

**Comment by flowerysong** [comment]
> I'm happy to go with your branch instead, you seem to know the backend better than I do anyway.

LOL. I don't even know enough Go to write "Hello, World!", I'm just good at faking it.

> If you want to pursue it, be aware that the ssh acceptance tests (enabled by setting VAULT_ACC env var) appear to have been broken by a recent change. Before we merge any PR, those should be fixed, and we should add tests for this change. I was planning on also setting up a docker container running openssh to run the acceptance tests so we can run them in CI. If you want to take that on, or you want to collaborate on these changes, let me know. I'll hold off on my work in the meantime.

I'm happy to let you carry on, I just wanted to offer my (somewhat tested) changes if you found them useful.

File: builtin/logical/ssh/path_sign.go (hunk 2 above)
Comment (ncabatoff) [reply builtin/logical/ssh/path_sign.go:530]: It's the signature length.  This is based on some a private method in the go crypto/ssh lib (certificateBytes).

File: builtin/logical/ssh/path_roles.go [hunk 4]
```diff
@@ -331,6 +333,16 @@ func pathRoles(b *backend) *framework.Path {
                                 If set, allows the enforcement of key types and minimum key sizes to be signed.
                                 `,
 			},
+			"algorithm_signer": &framework.FieldSchema{
```
Comment (ncabatoff) [inline builtin/logical/ssh/path_roles.go:336]: I'm implementing a default in the code, should make it explicit here instead.

File: builtin/logical/ssh/path_sign.go (hunk 1 above)
Comment (ncabatoff) [reply builtin/logical/ssh/path_sign.go]: Done.

File: builtin/logical/ssh/path_roles.go (hunk 3 above)
Comment (ncabatoff) [reply builtin/logical/ssh/path_roles.go]: Fixed.

File: builtin/logical/ssh/path_roles.go [hunk 5]
```diff
@@ -331,6 +333,17 @@ func pathRoles(b *backend) *framework.Path {
...
+				Type:    framework.TypeString,
+				Default: ssh.SigAlgoRSA,
+				Description: `
+				When supplied, this value specifies a signing algorithm for the key.  Possible values: 
+				ssh-rsa, rsa-sha2-256, rsa-sha2-512.
```
Comment (sgmiller) [inline builtin/logical/ssh/path_roles.go:341]: Probably needs a warning against ssh-rsa per our SHA1 warnings elsewhere, though possibly just in the docs.

File: builtin/logical/ssh/backend_test.go [hunk 6]
```diff
@@ -104,8 +104,95 @@ lObH9Faf0WGdnACZvTz22U9gWhw79S0SpDV31tC5Kl8dXHFiZ09vYUKkYmSd/kms
...
+		Env: []string{
+			"DOCKER_MODS=linuxserver/mods:openssh-server-openssh-client",
+			"PUBLIC_KEY=" + testPublicKeyInstall,
+			"SUDO_ACCESS=true",
+			"USER_NAME=vaultssh",
+		},
+		ExposedPorts: []string{"2222/tcp"},
+	})
+	if err != nil {
+		t.Fatalf("Could not start local MySQL docker container: %s", err)
```
Comment (vishalnayak) [inline builtin/logical/ssh/backend_test.go:142]: This error message probably needs updating.

File: builtin/logical/ssh/util.go [hunk 7]
```diff
@@ -102,8 +102,13 @@ func (b *backend) installPublicKeyInTarget(ctx context.Context, adminUser, usern
 	rmCmd := fmt.Sprintf("rm -f %s", scriptFileName)
 	targetCmd := fmt.Sprintf("%s;%s;%s", chmodCmd, scriptCmd, rmCmd)
 
-	session.Run(targetCmd)
-	return nil
+	//stderr := &bytes.Buffer{}
```
Comment (vishalnayak) [inline builtin/logical/ssh/util.go:105]: Is there a reason for the commented lines to exist?

**Pull request merged**